### ISSUE TITLE
LightGLue is slower than LoFTR

### ISSUE THREAD

**Issue by tcourat** [cvg/LightGlue#87]
Hi,

I am doing keypoint matching with 1280x720 images, with a GTX 1080 Ti ( 11.2 GiB of memory ) and found that LightGlue is slower than LoFTR, while it's supposed to be ~x4 faster according to the paper.

LoFTR takes ~0.45s to match, while LightGLue took 0.05s (superpoint)+0.63s (lightglue). There was ~900-1000 keypoints.

What could be the reason ? Even if the hardware is old it should not change the fact that LoFTR should be slower ?

**Comment by Phil26AT** [comment]
Hi, that should definitely not be the case. I suspect there is some overhead involved. Do you use compilation with LightGlue? If so, the compilation is done just-in-time, so when you call LightGlue the first time in your script it will take some time to compile, with the benefit of it being faster in the long run (assuming you run it over many pairs).

**Comment by ducha-aiki** [comment]
There is no way LightGlue is so slow, it usually takes like 0.02sec per 1024 points, unless you run on CPU (even though) or do the compilation.

**Comment by tcourat** [comment]
I am running on the GPU. I used the base demo notebook and changed no code (except for for the input images).
I ran the model several times and the matching time has the same order of magnitude (~0.5 or 0.6s).

Maybe my GPU is too old ?

**Comment by ducha-aiki** [comment]
What is your GPU then? And how exactly do you measure time? Could you please put full code snippet here?

**Comment by tcourat** [comment]
Hi, as I said earlier I have a  GTX 1080 Ti with 12 GB of VRAM.

I use the following code :

`from lightglue import LightGlue, SuperPoint, DISK, SIFT, ALIKED
from lightglue.utils import load_image, rbd
import time
import torch 

device = 'cuda' if torch.cuda.is_available() else 'cpu'

# SuperPoint+LightGlue
extractor = SuperPoint(max_num_keypoints=2048).eval().to(device)  # load the extractor
matcher = LightGlue(features='superpoint').eval().to(device) # load the matcher

# # or DISK+LightGlue, ALIKED+LightGlue or SIFT+LightGlue
# extractor = DISK(max_num_keypoints=2048).eval().cuda()  # load the extractor
# matcher = LightGlue(features='disk').eval().cuda()  # load the matcher

# load each image as a torch.Tensor on GPU with shape (3,H,W), normalized in [0,1]
image0 = load_image('./assets/sacre_coeur1.jpg').to(device)
image1 = load_image('./assets/sacre_coeur2.jpg').to(device)

# WARMING GPU - 10 matching
for i in range(10) : 
    # extract local features
    feats0 = extractor.extract(image0)  # auto-resize the image, disable with resize=None
    feats1 = extractor.extract(image1)

    # match the features
    matches01 = matcher({'image0': feats0, 'image1': feats1})

    feats0, feats1, matches01 = [rbd(x) for x in [feats0, feats1, matches01]]  # remove batch dimension
    matches = matches01['matches']  # indices with shape (K,2)
    points0 = feats0['keypoints'][matches[..., 0]]  # coordinates in image #0, shape (K,2)
    points1 = feats1['keypoints'][matches[..., 1]]  # coordinates in image #1, shape (K,2)

# Timing inference
start = torch.cuda.Event(enable_timing=True)
end = torch.cuda.Event(enable_timing=True)

start.record()
feats0 = extractor.extract(image0)  # auto-resize the image, disable with resize=None
feats1 = extractor.extract(image1)
matches01 = matcher({'image0': feats0, 'image1': feats1})
end.record()

# Waits for everything to finish running
torch.cuda.synchronize()

print(start.elapsed_time(end), 'ms')`

I know I should run it several time to average-out the inference time but it always end up with ~450ms +- 50ms with the following images.

**Comment by ducha-aiki** [comment]
It seems you don't turn the gradients off.
How about following:

```python3
torch.cuda.synchronize()
with torch.inference_mode():
    feats0 = extractor.extract(image0)  # auto-resize the image, disable with resize=None
    feats1 = extractor.extract(image1)
    matches01 = matcher({'image0': feats0, 'image1': feats1})
torch.cuda.synchronize()

```

**Comment by tcourat** [comment]
It is still roughly the same order of magnitude...

> It seems you don't turn the gradients off. How about following:
> 
> ```python
> torch.cuda.synchronize()
> with torch.inference_mode():
>     feats0 = extractor.extract(image0)  # auto-resize the image, disable with resize=None
>     feats1 = extractor.extract(image1)
>     matches01 = matcher({'image0': feats0, 'image1': feats1})
> torch.cuda.synchronize()
> ```

I added your advice :
start.record()
torch.cuda.synchronize()
with torch.inference_mode():
    feats0 = extractor.extract(image0)  # auto-resize the image, disable with resize=None
    feats1 = extractor.extract(image1)
    matches01 = matcher({'image0': feats0, 'image1': feats1})
# Waits for everything to finish running
torch.cuda.synchronize()
end.record()

But I still have the same order of magnitude of inference.

Note that I have also tried on a T4 GPU from Google COlab with the same snippet, and it only took ~110ms this time, but still far from the expected latency...

**Comment by hit2sjtu** [comment]
Tried your code on 1080Ti **without FlashAttention**, it only takes 96.ms
/home/***/code/LightGlue/lightglue/lightglue.py:141: UserWarning: FlashAttention is not available. For optimal speed, consider installing torch >= 2.0 or flash-attn.
  self.inner_attn = Attention(flash)
96.60598754882812 ms


Besides, I think you should exclude the timing of feature extraction when measuring LightGlue, since SuperPoint may be slower than expected.

After excluding feature extraction time, LG takes only 46.83689498901367 ms on 1080Ti.

```python

feats0 = extractor.extract(image0)  # auto-resize the image, disable with resize=None
feats1 = extractor.extract(image1)
torch.cuda.synchronize()
start.record()
with torch.inference_mode():
       matches01 = matcher({'image0': feats0, 'image1': feats1})

# Waits for everything to finish running
torch.cuda.synchronize()

end.record()


```

**Comment by hit2sjtu** [comment]
Using pytorch 2.0 slightly speed up inference on 1080Ti.
| Header | Header |
|--------|--------|
| With Extraction| Without Extraction|
| 96.6| 46.8 |
| 81.8| 41.9|

**Comment by tcourat** [comment]
Well then I do not know where the overhead is coming from...

**Comment by ducha-aiki** [comment]
@tcourat are you sure that your machine doesn't do any background process at the time of bechmarking?

**Comment by AnielAlexa** [comment]
Hy, do you have any updates on this? I'm facing the same issue on RTX 3070 also
